README.md
# Break The Blocks Game

> built using Three.js & Cannon.js

![playground](Images/README-files/main_menu.png)

<!-- adding space -->

## Contents

-   [Introduction](#introduction)
-   [Idea](#idea)
-   [Features](#features)
-   [ThreeX](#threex)
-   [game sounds](#game-sounds)
-   [animate function](#animate-function)
-   [Try it!](#try-it)
-   [ThreeJS](#threejs)
-   [CannonJS](#cannonjs)

---

<!-- adding space -->

## Introduction

In this Project I built the well-known game, Break The Blocks Game, built it using ThreeJS & CannonJS, with ZERO knowledge in ThreeJS & CannonJS, but still managed to learn these libraries and build the game.

It was really fun and challenging experience, I used really Interesting Technologies, Which are WebGL(Based on OpenGL) with the Help of ThreeJS library to abstract some of the work of creating 3D Objects and CannonJS to add Physics to the game.

you might not know anything about ThreeJS or CannonJS but this brief description hopefully gives you some idea about them, ThreeJS is popular library that is used to create and display animated 3D computer graphics in web browser using WebGL, the role of ThreeJS here is that, ThreeJS is doing you a favor, "the developer" by abstracting a lot of work and encapsulating many instructions in just one call to a method.

CannonJS, Which it's popular Physics library that can handle collisions and the movements of the objects inside the 3D World.

---

## Idea

Break The Block Game is well-known game that exists from 1970s, it's mechanism for playing is known, the player will try to deflect the bouncing ball so he/she break the blocks, once he/she break all the blocks, he/she will win, but if the ball bypass the bar then he/she will lose.

---

## Features

you will have two main components, first one is "the main menu", second one is "the playground", let us start with the main menu, in the main menu you will have the "title" of the game, "play" button and name of game creator going up and down.

you will be able to move the camera to any direction, in addition you can zoom in/out with the scroll wheel, in case if you zoomed out/in a bit much then the renderer will not render the menu 3D objects and you will not be able to click them, to fix that zoom in/out!, also you can enter full screen mode by double clicking anywhere on the screen.

the reason behind not seeing tha menu in case of over zooming out/in is
that it has something to do with the Camera frustum.

if you look at the code below you will see that we are giving 0.1 as value for the Camera frustum near plane and 1000 as value for Camera frustum far plane.

```
// PerspectiveCamera( fov : Number, aspect : Number, near : Number, far : Number )
const camera = new THREE.PerspectiveCamera(45, clientWidth / clientHeight, 0.1, 1000);
```

**here is a visual representation for this concept**
![Camera Frustum](/Images/README-files/VisualCameraFrustum.png)

**the renderer will not render any 3D objects outside the view frustum, 3D objects should be after near plane and before far plane, in aforementioned code
anything before 0.1 or after 1000 will not be visible, 0.1 and 1000 represents the distance from the original(0,0,0)(camera in the above picture or original in cartesian 3 dimensions plane) till the plane.**

**NOTE:** _ThreeJS support only vertical fov(fovY)_

**here is a visual representation for vertical fov or fovY**

![ThreeJS fovY](/Images/README-files/output-onlinepngtools.png)

in the playground you will have 52 colored blocks, that you have to destroy to win, you will have your purple bar that you can control using either the mouse or the keyboard keys/arrows, A & D keys or left & right arrows. (there a little hack here, you can exploit it to increase the speed of the bar, by pressing down together A and Left arrow or D and right arrow, the speed of the bar will double).

---

## ThreeX

Interaction is an essential part, without interacting with the 3D objects,
how would you play the game? here ThreeX comes to save us, ThreeX is a games extensions for ThreeJS, it includes many extensions that can help you, if you using ThreeJS to build games.

I chose Two extensions which are "ThreeX.domevents" and "ThreeX.keyboardState", from a long list of extensions, I will explain what are events and why we use them first and then I will go back to explain these two extensions.

when you want to add interactivity to a page to make it responsive you will certainly add events to these HTML elements, to know when this certain button got clicked or that 3D object in our case got clicked, using a method called addEventListener() of "document" object we can do that, but what if you want to attach events to ThreeJS 3D objects through Javascript? is it possible? apparently yes it's possible, using ThreeX.domevents and ThreeX.keyboardState extensions, will allow us interact with ThreeJS 3D objects in event-driven way,(by the way not all events are supported, the list of supported events for each extension is below).

you might wonder and ask yourself, what is then the difference between ThreeX.domevents and ThreeX.keyboardState?
ThreeX.domevents extension is used to add the following events to the 3D objects:

-   click
-   dblclick
-   mouseMove
-   mouseDown
-   mouseUp
-   touchmove
-   touchstart
-   touchend
-   contextmenu

assigning of the event will be like this:

```
domEvents.addEventListener(ThreeJS_Object, the_event, (the_event_object) => {
    the code...
});
```

you will notice that there is three parameters, first one "ThreeJS_Object" it's ThreeJS object that will be assigned "the_event", second parameter, "the_event" the type of the event you want to attach to this "ThreeJS_Object", third parameter is a function that gets an event object passed to it when it gets executed when the "the_event" happen, "the_event_object" will contain crucial information about the event that did happen.

so you will see that ThreeX.domevents doesn't actually consider keyboard events, so we need a solution to deal with keyboard pressed keys, fortunately there is another extension called "ThreeX.keyboardState" that can handle keyboard pressed keys elegantly with Keyboard.pressed('key') method that return true or false if the key is pressed or not, here is the list of the events it support:

-   Keydown
-   Keyup

```
Keyboard.pressed('a') // will return either true or false
```

---

## game sounds

![animation gif](Images/README-files/Animation.gif)

adding sounds to the game is an important thing to do, mixing the visual side with the auditory side is important, making the experience of the game more fun and exciting, so I added sound to the following:

-   the main menu
-   the playground
-   when clicking Play button
-   the ball hitting the blocks
-   in case of winning
-   in case of losing

---

## animate function

this is the most important function in this whole process, animate function gets registered so it can run every frame, it gets registered via requestAnimationFrame function, by passing it as an arugment to requestAnimationFrame.
if you have a screen that is 60FPS then this function gets executed 60 times per second, in other words your application gets rendered 60 times per second.

---

## Try it!
it's LIVE => [break-the-blocks-game](https://break-the-blocks-game.vercel.app)

---

## ThreeJS

**main target:** Three.js is a Javascript library that is used to create and display animated 3D objects in a web browser using WebGL.<br />
**contribution:** creating and animating 3D Objects.<br />
**citation:** https://threejs.org<br />
**license:** Three.js is licensed under MIT license, and it grants any person who obtains a copy of the software and associated files the right to use, copy, modify, merge, distribute, publish, sublicense and sell copies of the software.<br />
**alternatives:** BabylonJS<br />

---

## CannonJS

**main target:** A lightweight and simple 3D physics engine for the web.<br />
**contribution:** adding physics to the 3D World.<br />
**citation:** https://github.com/pmndrs/cannon-es<br />
**license:** Cannon.js is licensed under MIT license, and it grants any person who obtains a copy of the software and associated files the right to use, copy, modify, merge, distribute, publish, sublicense and sell copies of the software.<br />
**alternatives:** physijs<br />


CHANGELOG.md
#### 1.0.0 (2022-04-21)


src/js/mainMenu.mjs
import { playSound, mainMenuSound } from './loadingResources.mjs';
import '../public/css/main.css';
import '../public/css/github-corner.css';
import * as THREE from '../../threeJS/three.module.js';
import { THREEx } from '../../threeX/THREEx-Events/threex.domevents.mjs';
import KeyboardState from '../../threeX/THREEx-keyboard/threeX.keyboardState.js';
import { OrbitControls } from '../../threeJS/examples/jsm/controls/OrbitControls.js';
// prettier-ignore
import { title, play, creator, settings, mainMenuMusic, back, on, off, plusOn, minusOn } from './font.mjs';
import { particles } from './app/world/mainMenuStars.mjs';

// global variables
// Cancel Animation Variable
let cancel = null;

// y of on and off
let on_yAxis = 49;
let plusOn_yAxis = 50.5;
let minusOn_yAxis = 45.5;
let off_yAxis = 59;
let musicVolume = 0.1;
mainMenuSound.volume = musicVolume;

// making on button nonvisible
TweenMax.to(on.material, 0, { opacity: 0 });
TweenMax.to(plusOn.material, 0, { opacity: 0 });
TweenMax.to(minusOn.material, 0, { opacity: 0 });

// for keyboard events
const keyboard = new KeyboardState();

// scene
const scene = new THREE.Scene();
scene.background = new THREE.Color(0x000000);

// geometry
title.geometry.computeBoundingBox(); // computing the bounding box
title.geometry.center(); // centering the Mesh...
title.position.y = 105;

play.geometry.computeBoundingBox();
play.geometry.center();
play.position.y = 60;
play.rotation.x = -0.1;

settings.geometry.computeBoundingBox(); // computing the bounding box
settings.geometry.center(); // make it at the center
settings.position.y = 40;
settings.rotation.x = -0.1;

creator.geometry.computeBoundingBox(); // computing the bounding box
creator.geometry.center(); // make it at the center
creator.position.y = -40;
creator.position.z = -40;
creator.rotateX(-1.57 / 2 + 0.5); // in radian (1 radian == 57.3°)

mainMenuMusic.geometry.computeBoundingBox();
mainMenuMusic.geometry.center();
mainMenuMusic.position.y = -30000; // must return to 60
mainMenuMusic.rotation.x = -0.1;

back.geometry.computeBoundingBox(); // computing the bounding box
back.geometry.center(); // make it at the center
back.position.y = -30000; // must return to 40
back.rotation.x = -0.1;

on.geometry.computeBoundingBox(); // computing the bounding box
on.geometry.center(); // centering the Mesh...
on.position.y = -30000; // must return to 59
on.position.x = 28;
on.rotation.x = -0.1;
on.material.transparent = true;
plusOn.geometry.computeBoundingBox();
plusOn.geometry.center();
plusOn.position.x = 38;
plusOn.position.y = -30000;
plusOn.rotation.x = -0.1;
plusOn.material.transparent = true;
minusOn.geometry.computeBoundingBox();
minusOn.geometry.center();
minusOn.position.x = 38;
minusOn.position.y = -30000;
minusOn.rotation.x = -0.1;
minusOn.material.transparent = true;

off.geometry.computeBoundingBox(); // computing the bounding box
off.geometry.center(); // centering the Mesh...
off.position.y = -30000; // must return to 69
off.position.x = 31;
off.rotation.x = -0.1;
off.material.transparent = true;

// width, height
let clientWidth = document.documentElement.clientWidth;
let clientHeight = document.documentElement.clientHeight;

const renderer = new THREE.WebGLRenderer({
    antialias: true,
});

renderer.setSize(clientWidth, clientHeight);
document.body.append(renderer.domElement);

// camera
// prettier-ignore
const camera = new THREE.PerspectiveCamera(45, clientWidth / clientHeight, 0.1, 1000);

// THREEx
const domEvents = new THREEx.DomEvents(camera, renderer.domElement);

// OrbitControls
const controls = new OrbitControls(camera, renderer.domElement);
controls.enableDamping = true; // smooth stopping
controls.target = play.position.clone(); // to focus at the same as the play Object

camera.position.set(0, 100, 300);
controls.update();

// Helpers
// const AxesHelper = new THREE.AxesHelper(50);
// scene.add(AxesHelper);

// Clock
const clock = new THREE.Clock();

// adding to the scene
scene.add(
    camera,
    title,
    play,
    settings,
    creator,
    mainMenuMusic,
    back,
    on,
    plusOn,
    minusOn,
    off,
    particles
);

function animate() {
    cancel = requestAnimationFrame(animate);
    let ElapsedTime = clock.getElapsedTime(); //Get the seconds passed since the clock started

    // particles rotation
    particles.rotation.y += 0.001;

    // creator
    creator.position.y = -20 * Math.abs(Math.cos(ElapsedTime * 2));

    controls.update();
    renderer.render(scene, camera);
}

animate();

// events

document.addEventListener('visibilitychange', (event) => {
    if (document.visibilityState === 'visible' && on_yAxis === 59) {
        mainMenuSound.play();
    } else if (document.visibilityState === 'hidden' || on_yAxis === 49) {
        mainMenuSound.pause();
    }
});

window.addEventListener('resize', () => {
    clientWidth = document.documentElement.clientWidth;
    clientHeight = document.documentElement.clientHeight;
    camera.aspect = clientWidth / clientHeight;
    camera.updateProjectionMatrix();
    renderer.setSize(clientWidth, clientHeight);
});

document.addEventListener('keypress', (event) => {
    // prettier-ignore
    if (!document.fullscreenElement && (event.key === 'f' || event.key === 'F' )) {
        document.body.requestFullscreen();
        // prettier-ignore
    } else if(!document.fullscreenElement && (event.key === 'f' || event.key === 'F' )) {
        document.exitFullscreen();
    }
});

document.addEventListener('keydown', (event) => {
    if (event.key === 'ArrowUp') {
        // prettier-ignore
        mainMenuSound.volume = mainMenuSound.volume === 1.0 ? 1 : Number((mainMenuSound.volume + 0.1).toFixed(1));
    } else if (event.key === 'ArrowDown') {
        // prettier-ignore
        mainMenuSound.volume = mainMenuSound.volume === 0.0 ? 0 : Number((mainMenuSound.volume - 0.1).toFixed(1));
    }
});

domEvents.addEventListener(play, 'click', async (event) => {
    playSound.play();
    cancelAnimationFrame(cancel);
    // App render
    let { AppRenderer } = await import('./app/app.mjs');
    document.body.append(AppRenderer.domElement);
    renderer.domElement.remove();
});

domEvents.addEventListener(settings, 'click', function (event) {
    // prettier-ignore
    {
        gsap.to(title.position, { duration: 10.5, ease: "bounce.out", y: 30000 , overwrite: true });
        gsap.to(play.position, { duration: 10.5, ease: "bounce.out", y: 30000, overwrite: true });
        gsap.to(settings.position, { duration: 10.5, ease: "bounce.out", y: 30000, overwrite: true });
        gsap.to(creator.position, { duration: 10.5, ease: "bounce.out", x: -30000, overwrite: true });
        gsap.to(mainMenuMusic.position, { duration: 2, ease: "power4.out", y: 60 , overwrite: true });
        gsap.to(back.position, { duration: 2, ease: "power4.out", y: 40 , overwrite: true });
        gsap.to(on.position, { duration: 2, ease: "power4.out", y: on_yAxis , overwrite: true });
        gsap.to(plusOn.position, { duration: 2, ease: "power4.out", y: plusOn_yAxis , overwrite: true });
        gsap.to(minusOn.position, { duration: 2, ease: "power4.out", y: minusOn_yAxis , overwrite: true });
        gsap.to(off.position, { duration: 2, ease: "power4.out", y: off_yAxis, overwrite: true });
    }
});

domEvents.addEventListener(back, 'click', function (event) {
    // prettier-ignore
    {
        gsap.to(mainMenuMusic.position, { duration: 10.5, ease: "bounce.out", y: -30000, overwrite: true });
        gsap.to(back.position, { duration: 10.5, ease: "bounce.out", y: -30000, overwrite: true });
        gsap.to(on.position, { duration: 10.5, ease: "bounce.out", y: -30000, overwrite: true });
        gsap.to(plusOn.position, { duration: 10.5, ease: "bounce.out", y: -30000, overwrite: true });
        gsap.to(minusOn.position, { duration: 10.5, ease: "bounce.out", y: -30000, overwrite: true });
        gsap.to(off.position, { duration: 10.5, ease: "bounce.out", y: -30000, overwrite: true });
        gsap.to(title.position, { duration: 2, ease: "power4.out", y: 105, overwrite: true });
        gsap.to(play.position, { duration: 2, ease: "power4.out", y: 60, overwrite: true });
        gsap.to(settings.position, { duration: 2, ease: "power4.out", y: 40, overwrite: true });
        gsap.to(creator.position, { duration: 2, ease: "power4.out", x: 0, overwrite: true });
    }
});

function plusOnButtonFunction() {
    // prettier-ignore
    mainMenuSound.volume = mainMenuSound.volume === 1.0 ? 1 : Number((mainMenuSound.volume + 0.1).toFixed(1));
}

function minusOnButtonFunction() {
    // prettier-ignore
    mainMenuSound.volume = mainMenuSound.volume === 0.1 ? 0.1 : Number((mainMenuSound.volume - 0.1).toFixed(1));
}

function onButtonFunction() {
    // prettier-ignore
    {
        domEvents.unbind(on, 'click', onButtonFunction);
        domEvents.unbind(plusOn, 'click', plusOnButtonFunction);
        domEvents.unbind(minusOn, 'click', minusOnButtonFunction);
        domEvents.bind(off, 'click', offButtonFunction);
        mainMenuSound.pause();
        gsap.fromTo(on.position, {}, { y: (on_yAxis -= 10), duration: 1 });
        gsap.fromTo(plusOn.position, {}, { y: (plusOn_yAxis -= 10), duration: 1 });
        gsap.fromTo(minusOn.position, {}, { y: (minusOn_yAxis -= 10), duration: 1 });
        gsap.fromTo(off.position, {}, { y: (off_yAxis -= 10), duration: 1 });
        TweenMax.to(on.material, 1, { opacity: 0 });
        TweenMax.to(plusOn.material, 1, { opacity: 0 });
        TweenMax.to(minusOn.material, 1, { opacity: 0 });
        TweenMax.to(off.material, 1, { opacity: 1 });
    }
}

function offButtonFunction() {
    // prettier-ignore
    {
        domEvents.unbind(off, 'click', offButtonFunction);
        domEvents.bind(on, 'click', onButtonFunction);
        domEvents.bind(plusOn, 'click', plusOnButtonFunction);
        domEvents.bind(minusOn, 'click', minusOnButtonFunction);
        mainMenuSound.play();
        gsap.fromTo(on.position, {}, { y: (on_yAxis += 10), duration: 1 });
        gsap.fromTo(plusOn.position, {}, { y: (plusOn_yAxis += 10), duration: 1 });
        gsap.fromTo(minusOn.position, {}, { y: (minusOn_yAxis += 10), duration: 1 });
        gsap.fromTo(off.position, {}, { y: (off_yAxis += 10), duration: 1 });
        // in 1 second it goes to opacity=1
        TweenMax.to(on.material, 1, { opacity: 1 });
        TweenMax.to(plusOn.material, 1, { opacity: 1 });
        TweenMax.to(minusOn.material, 1, { opacity: 1 });
        TweenMax.to(off.material, 1, { opacity: 0 });
    }
}

// prettier-ignore
domEvents.bind(off, 'click', offButtonFunction);

// let timer = null;

// domEvents.addEventListener(exit, 'click', (event) => {
//     exitSound.play();
//     document.querySelector('#overlay').style.display = 'flex';
//     let sec = 3;
//     document.getElementById('timer').innerHTML = '😢';
//     timer = setInterval(function () {
//         document.getElementById('timer').innerHTML = '' + sec;
//         sec--;
//         if (sec < 0) {
//             clearInterval(timer);
//             cancelAnimationFrame(cancel);
//         }
//     }, 1000);
// });

// window.off = function () {
//     clearInterval(timer);
//     document.querySelector('#overlay').style.display = 'none';
// };

window.playAgain = function () {
    window.open('https://break-the-blocks-game.vercel.app', '_self');
};
